source/tests/system/nidmm_driver_api_tests.cpp

```diff
@@ -0,0 +1,282 @@
+        }
+
+    private:
+        std::shared_ptr<::grpc::Channel> channel_;
+        std::unique_ptr<::nidevice_grpc::Session> driver_session_;
+        std::unique_ptr<dmm::NiDMM::Stub> nidmm_stub_;
+        std::unique_ptr<nidevice_grpc::SessionRepository> session_repository_;
+        std::unique_ptr<dmm::NiDMMLibrary> nidmm_library_;
+        std::unique_ptr<dmm::NiDMMService> nidmm_service_;
+        std::unique_ptr<::grpc::Server> server_;
+};
+
+TEST_F(NiDMMDriverApiTest, NiDMMSelfTest_SendRequest_SelfTestCompletesSuccessfully)
+{
+  ::grpc::ClientContext context;
+  dmm::SelfTestRequest request;
+  request.mutable_vi()->set_id(GetSessionId());
+  dmm::SelfTestResponse response;
+
+  ::grpc::Status status = GetStub()->SelfTest(&context, request, &response);
+
+  EXPECT_TRUE(status.ok());
+  expect_api_success(response.status());
+  EXPECT_EQ(0, response.self_test_result());
+  EXPECT_LT(0, strlen(response.self_test_message().c_str()));
+}
+
+TEST_F(NiDMMDriverApiTest, NiDMMReset_SendRequest_ResetCompletesSuccessfully)
+{
+  ::grpc::ClientContext context;
+  dmm::ResetRequest request;
+  request.mutable_vi()->set_id(GetSessionId());
+  dmm::ResetResponse response;
+
+  ::grpc::Status status = GetStub()->Reset(&context, request, &response);
+
+  EXPECT_TRUE(status.ok());
+  expect_api_success(response.status());
+}
+
+TEST_F(NiDMMDriverApiTest, NiDMMSetViReal64Attribute_SendRequest_GetViReal64AttributeMatches)
+{
+  const char* channel_name = "";
+  const dmm::NiDMMAttributes attribute_to_set = dmm::NiDMMAttributes::NIDMM_ATTRIBUTE_TRIGGER_DELAY;
+  const ViReal64 expected_value = 42.24;
+  ::grpc::ClientContext context;
+  dmm::SetAttributeViReal64Request request;
+  request.mutable_vi()->set_id(GetSessionId());
+  request.set_channel_name(channel_name);
+  request.set_attribute_id(attribute_to_set);
+  request.set_attribute_value(expected_value);
+  dmm::SetAttributeViReal64Response response;
+
+  ::grpc::Status status = GetStub()->SetAttributeViReal64(&context, request, &response);
+  EXPECT_TRUE(status.ok());
+  expect_api_success(response.status());
+  
+  ViReal64 get_attribute_value = get_real64_attribute(channel_name, attribute_to_set);
+  EXPECT_EQ(expected_value, get_attribute_value);
+}
+
+TEST_F(NiDMMDriverApiTest, NiDMMSetViInt32Attribute_SendRequest_GetViInt32AttributeMatches)
+{
+  const char* channel_name = "";
+  const dmm::NiDMMAttributes attribute_to_set = dmm::NiDMMAttributes::NIDMM_ATTRIBUTE_SAMPLE_COUNT;
+  const ViInt32 expected_value = 4;
+  ::grpc::ClientContext context;
+  dmm::SetAttributeViInt32Request request;
+  request.mutable_vi()->set_id(GetSessionId());
+  request.set_channel_name(channel_name);
+  request.set_attribute_id(attribute_to_set);
+  request.set_attribute_value(expected_value);
+  dmm::SetAttributeViInt32Response response;
+
+  ::grpc::Status status = GetStub()->SetAttributeViInt32(&context, request, &response);
+  EXPECT_TRUE(status.ok());
+  expect_api_success(response.status());
```

Comment:
This belongs with the other validation logic.

Reply:
moved it with the block below